RESULT_PROPOSAL.md

```diff
@@ -0,0 +1,295 @@
+}
+```
+
+### Java
+
+```java
+PayPalLauncher payPalLauncher = new PayPalLauncher(paymentAuthResult ->
+        payPalClient.tokenize(paymentAuthResult, (paymentMethodNonce, error) -> {
+            if (error != null) {
+                if (error instanceof UserCanceledException) {
+                    // handle user canceled payment flow
+                } else {
+                    // handle error 
+                }
+                return;
+            }
+            if (paymentMethodNonce != null) {
+                // send nonce to server
+                return;
+            }
+            // unexpected error - one of nonce or error should not be null
+        }));
+
+payPalClient.createPaymentAuthRequest(activity, request, (paymentAuthRequest, error) -> {
+    if (error != null) {
+        // handle error
+        return;
+    }
+    if (paymentAuthRequest != null) {
+        payPalLauncher.launch(activity, paymentAuthRequest);
+        return;
+    }
+    // unexpected error - one of paymentAuthRequest or error should not be null
+});
+```
+
+## Listener
+
+This integration aligns most closely with the v4 integration listener pattern, but either
+requires two separate listeners (for launcher and client), or couples the launcher and the
+client because both need to be invoked in different listener methods. The benefit of a single
+listener is that errors throughout the entire flow can be handled in one method (rather than two
+separate error handling points with the callback patterns).
+
+### Kotlin
+
+```kotlin
+payPalLauncher = PayPalLauncher(listener) 
+payPalClient = PayPalClient(listener)
+payPalClient.createPaymentAuthRequest(activity, request)
+
+override fun onPaymentAuthRequest(paymentAuthRequest: PaymentAuthRequest) {
+    payPalLauncher.launch(paymentAuthRequest)
+}
+
+override fun onPaymentAuthResult(paymentAuthResult: PaymentAuthResult) {
+    payPalClient.tokenize(paymentAuthResult)
+}
+
+override fun onSuccess(payPalAccountNonce: PayPalAccountNonce) {
+    // send nonce to server
+}
+
+override fun onCancel() {
+    // handle cancel
+}
+
+override fun onError(error: Exception) {
+    // handle error
+}
+```
+
+### Java
+
+```java
+PayPalLauncher payPalLauncher = new PayPalLauncher(listener);
+PayPalClient payPalClient = new PayPalClient(listener);
+payPalClient.createPaymentAuthRequest(activity, request);
+
+@Override
+public void onPaymentAuthRequest(PaymentAuthRequest paymentAuthRequest) {
+    payPalLauncher.launch(paymentAuthRequest)
+}
+
+@Override
+public void onPaymentAuthResult(PaymentAuthResult paymentAuthResult) {
+    payPalClient.tokenize(paymentAuthResult)
+}
+
+@Override
+public void onSuccess(PayPalAccountNonce payPalAccountNonce) {
+    // send nonce to server
+}
+
+@Override
+public void onCancel() {
+    // handle cancel
+}
+
+@Ovrride
+public void onError(Exception error) {
+    // handle error
+}
+```
+## Single Result Object With Types
```

Comment:
This _feels_ the best to me, but it also reads similar to Swift so I think that makes sense. It sounds like this is also what other folks have moved to so would be a good move to modernize our approach.

Is there a world in which we'd consider offering 2 similar approaches in v5 and in a future v6 go with 1? In iOS v6 we offer both [completions](https://github.com/braintree/braintree_ios/blob/b0db748229adf147f53573ea2a567ee6d74f43de/Sources/BraintreePayPal/BTPayPalClient.swift#L70-L75) and [async/await](https://github.com/braintree/braintree_ios/blob/b0db748229adf147f53573ea2a567ee6d74f43de/Sources/BraintreePayPal/BTPayPalClient.swift#L88-L98) with a plan to go to only async/await in v7. I am not sure if something like this is feasible on Android though or would over complicate things.

Reply:
It would be a bit of a docs and development burden, but I could see the case for including the single result callback pattern and the listener pattern (since listener was the suggested v4 approach) to ease integration burden and confirm which approach merchants prefer. But that might create a whole mess of docs (you'd need to invoke different methods if you set the listener vs not)

Reply:
Ah gotcha - that makes sense. We just document 1 for iOS but we can kinda just call the callbacks from the async/await so it's not a ton of extra code lift. Whereas in Android it seems like we'd be adding a ton of extra work for ourselves.

Reply:
@sarahkoop - After our mob PR review, I also vote for this option. We _can_ take a look at the past few inbounds in Slack & GH to see if they are mostly Kotlin (probably). 

Historically we've been more conservative, but I think we're at a point now we can be more aggressive and go for this. We could also introduce it module-by-module and put out beta releases to see if people use it and have feedback.